text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,159 @@
+The rest of this section just illustrate hows they would likely be implemented.
+
+## `Cargo.toml` Interface
+
+First and foremost, one will now be able to depend on standard library crates by version, e.g. `std = "1.10"`.
+This will work just as if `std` was on crates.io---features and other modifiers are supported.
+
+For backwards compatibility, Cargo must inject such standard library dependencies for existing packages.
+Exactly which dependencies is unresolved, but a requirement at least as strong as `std = "^1.0"` as a primary and build dependency is assured.
+
+Now, not all crates depend on `std`, so there must be a way to opt out.
+For this, we introduce a new `implicit-dependencies` key.
+It is defined by default as:
+```toml
+implicit-dependencies = ["primary", "build", "dev"]
+```
+This indicates each of `dependencies`, `build-dependencies`, and `dev-dependencies` maps (respectively) is augmented with implicit elements.
+A manual definition may be that or almost any subset, in which case only the included dependency maps are augmented.
+The one additional rule is `"dev"` must be included in the set: we have no plan for Cargoizing the default test runner (either the attribute syntax manipulation or runtime), and wish to be forward-compatible with doing so in the future.
+
+Finally, if an (explicit) dependency conflicts with one of the implicit defaults, that category of implicit dependency will be skipped.
+For example, if a crate explicit depends on `std` as a build-dependency, neither `std` nor any other implicit build dependency will be injected.
+
+## Cargo Command Line Interface
+
+A flag will be added
+```
+--resolve-compiler-specific=bin,src
+```
+where either `bin` or `src` may be included in the set.
+If `bin` is included, Cargo will allow the use of sysroot binaries to satisfy deps.
+It is assumed that the version of all sysroot rlibs is the same as the version of Rust which the compiler implements.
+If `src` is included, Cargo will fallback on a compiler-provided package registry when other registries (e.g. crates.io) fail to provide a package.
+For backwards compatibility, the default is `--resolve-compiler-specific=bin`.
+
+## Compiler source packaging
+
+While the standard library *interface* defined with each rustc version, the implementation, by virtue of using unstable features, is compiler-specific.
+This makes the standard library unfit for crates.io.
+(Additionally, the issue of dealing with nightly also makes crates.io hard to use, but that is a less clear-cut obstacle.)
+
+Cargo will soon gain the ability to create and chain custom registries, as described in
+https://github.com/rust-lang/cargo/pull/2361 .
+Compiler's should package the source of their implementation of the standard library as a registry, which can be distributed with the compiler.
+In practice, rustc will optionally contain the source in its sysroot.
+Rustup may be able to put it there if the default download does not contain it already.
```

Comment:
So the dirty secret is that my goal of trying to keep Cargo from thinking about the stdlib for regularity's sake, I've made writing such an error rather difficult. E.g. implicit deps (which are just a subset of the stdlib anyways) are injected at the beginning of the pipeline, so most of Cargo won't know what an implicit dep is. The "compiler registry" and "mock registry" (the one for the sysroot binaries, it probably needs a better name) would be appended to this fallback chain whose purpose is broader than this (e.g. big company disabling crates.io in your example from the registry PR).

I did think of a solution however that largely allows keeping the desugaring implementation: whenever Cargo encounters *any* missing package, it checks to see if a binary is in the sysroot. If so, it can print a message like "foo with version matching requirement bar could not be found, but we see an rlib with the same name that came with the compiler. Perhaps you should do xyz."